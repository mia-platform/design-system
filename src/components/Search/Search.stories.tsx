/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'
import { FaDiamond } from 'react-icons/fa6'
import { Flex } from 'antd'

import { Icon } from '../Icon'
import { Search } from '.'
import { Typography } from '../Typography'

const meta = {
  component: Search,
  args: {
    placeholder: 'Placeholder...',
  },
  argTypes: {
  },
} satisfies Meta<typeof Search>

export default meta
type Story = StoryObj<typeof meta>

const options = [
  ...Array(5).keys(),
].map((id) => ({
  value: `value ${id + 1}`,
}))

export const Default: Story = {
  args: {
    options,
    defaultValue: 'Default',
  },
}

export const Borderless: Story = {
  args: {
    options,
    appearance: Search.Appearance.Borderless,
    defaultValue: 'Borderless',
  },
}

export const ReadOnly: Story = {
  args: {
    options,
    isReadOnly: true,
    defaultValue: 'Read-only',
  },
}

export const Error: Story = {
  args: {
    options,
    isError: true,
    defaultValue: 'Error',
  },
}

export const Disabled: Story = {
  args: {
    options,
    isDisabled: true,
    defaultValue: 'Disabled',
  },
}

export const AllowClear: Story = {
  args: {
    options,
    allowClear: true,
    defaultValue: 'Allow clear',
  },
}

export const CustomLabels: Story = {
  args: {
    options: options.map(({ value }) => ({
      value,
      label: (
        <Flex align="center" gap={8}>
          <Icon component={FaDiamond} size={16} />
          <Typography.BodyM isBold>{value}</Typography.BodyM>
        </Flex>
      ),
    })),
    defaultValue: 'Custom labels',
  },
}
