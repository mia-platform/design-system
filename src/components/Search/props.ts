/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { BaseInputProps } from '../BaseInput/props'
import { IconComponent } from '../Icon/Icon.props'

export type SearchItem<ValueType> = {
  value: ValueType
  label?: ReactNode
}

export type SearchChangeHandler<ValueType> = (
  value: ValueType,
  option: SearchItem<ValueType> | SearchItem<ValueType>[]
) => void

export type SearchSelectHandler<ValueType> = (
  value: ValueType extends (infer E)[] ? E : ValueType,
  option: SearchItem<ValueType>
) => void

export type SearchProps<ValueType> = BaseInputProps & {

  /**
   * The input content value.
   */
  value?: ValueType

  /**
   * The default valute for the input.
   */
  defaultValue?: ValueType

  /**
  * Select options.
   */
  options: SearchItem<ValueType>[]

  /**
   * The left icon for the Input.
   */
  iconLeft?: IconComponent

  /**
   * The right icon for the Input.
   */
  iconRight?: IconComponent

  /**
   * If allow to remove input content with clear icon.
   */
  allowClear?: boolean

  /**
   * Callback when user input.
   */
  onChange?: SearchChangeHandler<ValueType>

  /**
   * Callback when an option is selected.
   */
  onSelect?: SearchSelectHandler<ValueType>

  /**
   * Callback when searching.
   */
  onSearch?: (value: string) => void
}
