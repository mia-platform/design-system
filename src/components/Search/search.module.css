
.search {
  & :global(.mia-platform-input-suffix) {
    color: var(--palette-text-neutral-main, #636363) !important;
  }
  & input[disabled] ~ :global(.mia-platform-input-suffix) {
    color: var(--palette-text-neutral-subtler, #ACACAC) !important;
  }
}
