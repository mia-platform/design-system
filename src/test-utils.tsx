/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
