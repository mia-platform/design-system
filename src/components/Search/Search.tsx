/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { AutoComplete as AntAutoComplete, AutoCompleteProps as AntAutoCompleteProps, InputRef } from 'antd'
import { ReactElement, Ref, forwardRef as forwardref, useCallback } from 'react'
import { PiMagnifyingGlass } from 'react-icons/pi'

import { BaseInput, defaults as baseInputDefaults } from '../BaseInput/BaseInput'
import { SearchItem, SearchProps } from './props'
import { Appearance } from '../BaseInput/types'
import { Dropdown } from '../Dropdown'
import { Input } from '../Input/index'
import { InputProps } from '../Input/props'
import styles from './search.module.css'

/**
 * A UI element to insert long text content in a form.
 *
 * @link https://ant.design/components/Autocomplete
 * @returns {Search} Search component
 */

export const defaults = {
  ...baseInputDefaults,
  iconRight: PiMagnifyingGlass,
}

const InnerInput = forwardref(
  function InnerInput(props: InputProps, ref: Ref<InputRef>): ReactElement {
    return <Input inputRef={ref} type={Input.Type.Search} {...props} />
  }
)

export const Search = <ValueType, >(
  {
    appearance = defaults.type,
    value,
    defaultValue,
    isDisabled,
    isReadOnly,
    isFullWidth = defaults.isFullWidth,
    isError,
    placeholder,
    allowClear,
    options,
    inputRef,
    iconLeft,
    iconRight = defaults.iconRight,
    onChange,
    onSelect,
    onSearch,
    dropdownProps,
  }: SearchProps<ValueType>) : ReactElement => {
  const dropdownRender = useCallback((menu: ReactElement) => {
    const props = { children: menu, items: [], ...dropdownProps }
    return <Dropdown {...props} />
  }, [dropdownProps])

  return (
    <BaseInput <AntAutoCompleteProps<ValueType, SearchItem<ValueType>>>
      appearance={appearance}
      className={styles.search}
      component={AntAutoComplete}
      defaultValue={defaultValue}
      dropdownRender={dropdownRender}
      inputRef={inputRef}
      isDisabled={isDisabled}
      isError={isError}
      isFullWidth={isFullWidth}
      isReadOnly={isReadOnly}
      options={options}
      placeholder={placeholder}
      value={value}
      onChange={onChange}
      onSearch={onSearch}
      onSelect={onSelect}
    >
      <InnerInput
        allowClear={allowClear}
        appearance={appearance}
        iconLeft={iconLeft}
        iconRight={iconRight}
        isDisabled={isDisabled}
        isError={isError}
        isFullWidth={isFullWidth}
        isReadOnly={isReadOnly}
      />
    </BaseInput>
  )
}

Search.Appearance = Appearance
