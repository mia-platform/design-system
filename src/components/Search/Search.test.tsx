/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render, screen } from '../../test-utils'
import { Search } from './Search'

const options = [
  ...Array(5).keys(),
].map((id) => ({
  value: `value ${id + 1}`,
}))

describe('Input Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders correctly', () => {
    const { asFragment } = render(<Search options={options} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Should call onChange handler on typing', async() => {
    const onChange = jest.fn()
    const text = 'text'

    render(<Search options={options} onChange={onChange} />)

    const input = screen.getByRole<HTMLInputElement>('searchbox')

    fireEvent.change(input, { target: { value: text } })

    expect(onChange).toHaveBeenCalledWith(text, {})
    expect(input.value).toBe(text)
  })

  test('Should onSearch handler on typing', async() => {
    const onSearch = jest.fn()
    const text = 'text'

    render(<Search options={options} onChange={onSearch} />)

    const input = screen.getByRole<HTMLInputElement>('searchbox')

    fireEvent.change(input, { target: { value: text } })

    expect(onSearch).toHaveBeenCalledWith(text, {})
    expect(input.value).toBe(text)
  })

  test('Allow clear should clear input', () => {
    const [{ value: defaultValue }] = options

    render(<Search allowClear defaultValue={defaultValue} options={options} />)

    const input = screen.getByRole<HTMLInputElement>('searchbox')
    const clearIcon = screen.getByRole('img', { name: 'close-circle' })

    fireEvent.click(clearIcon)

    expect(input.value).toBe('')
  })

  test('Should show the options on typing', () => {
    render(<Search allowClear options={options} />)

    const input = screen.getByRole<HTMLInputElement>('searchbox')

    fireEvent.change(input, { target: { value: 'value' } })

    options.forEach(({ value }) => {
      expect(screen.getByTitle(value)).toBeInTheDocument()
    })
  })

  test('Pressing an option should trigger onSelect', () => {
    const onSelect = jest.fn()

    render(<Search allowClear options={options} onSelect={onSelect} />)

    const input = screen.getByRole<HTMLInputElement>('searchbox')

    fireEvent.change(input, { target: { value: 'value' } })

    const option = screen.getByTitle('value 1')

    fireEvent.click(option)

    expect(onSelect).toHaveBeenCalledWith(options[0].value, options[0])
  })
})
