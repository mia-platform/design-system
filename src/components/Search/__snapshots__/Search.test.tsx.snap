// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="mia-platform-select mia-platform-select-outlined mia-platform-select-auto-complete input fullWidth search mia-platform-select-single mia-platform-select-customize-input mia-platform-select-show-search"
  >
    <div
      class="mia-platform-select-selector"
    >
      <span
        class="mia-platform-select-selection-search"
      >
        <span
          class="mia-platform-input-affix-wrapper mia-platform-input-outlined input fullWidth input"
        >
          <input
            class="mia-platform-input"
            type="search"
            value=""
          />
          <span
            class="mia-platform-input-suffix"
          >
            <svg
              color="currentColor"
              fill="currentColor"
              height="12"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M229.66,218.34l-50.07-50.06a88.11,88.11,0,1,0-11.31,11.31l50.06,50.07a8,8,0,0,0,11.32-11.32ZM40,112a72,72,0,1,1,72,72A72.08,72.08,0,0,1,40,112Z"
              />
            </svg>
          </span>
        </span>
      </span>
      <span
        class="mia-platform-select-selection-placeholder"
      />
    </div>
  </div>
</DocumentFragment>
`;
